# Day 3: Functions & File I/O

1. Overview  
2. Defining & Calling Functions  
3. Variable Scope & Return Values  
4. File Input/Output (CSV & Text)  
5. Exercise 1: Solar CSV Parser Function  
6. Exercise 2: Demand Stats Function  
7. Exercise 3: Generate & Write Reports  

---

## 1. Overview

Welcome to Day 3 of my Energy Analytics journey! Today’s goals are to:

- Define and use Python functions to encapsulate reusable logic  
- Understand variable scope and return values  
- Read from and write to CSV and text files  
- Complete three field‑related coding exercises in Jupyter notebook  

---

## 2. Defining & Calling Functions

In this section, I will:

- Use the `def` keyword to define a function  
- Pass positional and keyword arguments, including default values  
- Return values vs. printing inside the function  
- Write docstrings for clarity  

In [1]:
def convert_wh_to_kwh(wh):
    """Convert a Wh value to kWh."""
    return wh / 1000

print(convert_wh_to_kwh(5200))

5.2


---

## 3. Variable Scope & Return Values

I will note:

- **Local scope**: variables defined inside a function  
- **Global scope**: variables defined outside a function  
- How `return` ends my function execution and hands back a value  
- Best practices: avoid modifying globals inside functions  

---

## 4. File Input/Output (CSV & Text)

In this section, I will:

- Open files using `open(path, mode)` with modes `r`, `w`, `a`  
- Use `with` statements for safe file handling  
- Read and write plain text (`.read()`, `.write()`)  
- Use the `csv` module or `pandas` for CSV I/O  

In [2]:
# Text I/O
with open("sample.txt", "w") as f:
    f.write("Energy Analytics Day 3\n")
with open("sample.txt", "r") as f:
    print(f.read())

Energy Analytics Day 3



In [3]:
# CSV I/O via csv module
import csv
rows = [["day","irradiance_wh"], [1,5200], [2,4800]]
with open("irr.csv", "w", newline="") as f:
    writer = csv.writer(f)
    writer.writerows(rows)
with open("irr.csv", "r") as f:
    for row in csv.reader(f):
        print(row)

['day', 'irradiance_wh']
['1', '5200']
['2', '4800']


---

### Exercise 1: Solar CSV Parser Function

In this cell, I create `parse_solar_csv(path)` that:

- Opens `irr.csv` (with columns `day` and `irradiance_wh`)  
- Converts each Wh/m² value to kWh/m² by dividing by 1,000  
- Returns a list of kWh values  

- **Why?**  
  I need to standardize my raw irradiance data for downstream analysis.  
- **How?**  
  I use `csv.DictReader` inside a `with open(..., "r")` block and append converted values to a list.  

In [1]:
import csv

def parse_solar_csv(path) :
    """Read csv and convert Wh to KWh list."""
    kwh = []
    with open(path, "r") as f :
        reader = csv.DictReader(f)
        for row in reader :
            kwh.append(float(row["irradiance_wh"]) / 1000)
    return kwh

print(parse_solar_csv("irr.csv"))

[5.2, 4.8, 5.3, 5.1, 4.95]


---

### Exercise 2: Demand Stats Function

In this cell, I create `demand_stats(demands)` that:

- Accepts a list of my daily peak demand values (GW)  
- Computes `min`, `max`, and average  
- Returns a tuple `(min, max, avg)`  

- **Why?**  
  Summarizing my demand statistics is essential for understanding grid load variability.  
- **How?**  
  I use built‑in functions `min()`, `max()`, and `sum()/len()` inside my function and return the results.  

In [2]:
def demand_stats(demants) :
    """Return (min, max, avg) of demands list."""
    return (min(demants), max(demants), sum(demants)/len(demants))

stats = demand_stats([35.2, 42.8, 56.5, 64.2])
print(stats)

(35.2, 64.2, 49.675)


---

### Exercise 3: Generate & Write Reports

In this cell, I:

1. Call `parse_solar_csv("irr.csv")` → `solar_kwh` list  
2. Call `demand_stats(peak_demands)` on  
   `[35.2, 42.8, 61.5, 55.0, 38.9]` → `stats` tuple  
3. Open (or create) `report.txt` in write mode and write lines:                                                                                 
   Solar kwh days: [5.2, 4.8, 5.3, 5.1, 4.95]                                                                                                 
   Demand stats : min=35.2, max=64.2, avg=51.48


- **Why?**  
Automated reporting ensures my results are easily shared and reproducible.  
- **How?**  
I use `with open("report.txt", "w")` and `f.write()` for each line.  

In [3]:
solar_kwh = parse_solar_csv("irr.csv")
stats = demand_stats([35.2, 42.8, 56.5, 64.2, 58.7])

with open("report.txt", "w") as f :
    f.write(f"Solar kwh days: {solar_kwh}\n")
    f.write(f"Demand stats : min={stats[0]}, max={stats[1]}, avg={stats[2]:.2f}\n")